apps/src/sites/studio/pages/components/StageCard.jsx

```diff
@@ -0,0 +1,166 @@
+    color: '#5b6770',
+    marginBottom: 15
+  },
+  stageLockable: {
+    fontSize: 13,
+    marginTop: 3
+  },
+  addLevel: {
+    fontSize: 14,
+    background: '#eee',
+    border: '1px solid #ddd',
+    boxShadow: 'inset 0 1px 0 0 rgba(255, 255, 255, 0.8)',
+    margin: '5px 0'
+  }
+};
+
+const StageCard = React.createClass({
+  propTypes: {
+    reorderLevel: React.PropTypes.func.isRequired,
+    addLevel: React.PropTypes.func.isRequired,
+    stagesCount: React.PropTypes.number.isRequired,
+    stage: React.PropTypes.object.isRequired
+  },
+
+  getInitialState() {
+    return {currentPositions: []};
+  },
+
+  metrics(token) {
+    return ReactDOM.findDOMNode(this.refs[`levelToken${token}`]).getBoundingClientRect();
+  },
+
+  handleDragStart(position, {pageY}) {
+    const startingPositions = this.props.stage.levels.map(level => {
+      const metrics = this.metrics(level.position);
+      return metrics.top + metrics.height / 2;
+    });
+    this.setState({
+      drag: position,
+      dragHeight: this.metrics(position).height + levelTokenMargin,
+      initialPageY: pageY,
+      initialScroll: document.body.scrollTop,
+      newPosition: position,
+      startingPositions
+    });
+    window.addEventListener('selectstart', this.preventSelect);
+    window.addEventListener('mousemove', this.handleDrag);
+    window.addEventListener('mouseup', this.handleDragStop);
+  },
+
+  handleDrag({pageY}) {
+    const scrollDelta = document.body.scrollTop - this.state.initialScroll;
+    const delta = pageY - this.state.initialPageY;
+    const dragPosition = this.metrics(this.state.drag).top + scrollDelta;
+    let newPosition = this.state.drag;
+    const currentPositions = this.state.startingPositions.map((midpoint, index) => {
+      const position = index + 1;
+      if (position === this.state.drag) {
+        return delta;
+      }
+      if (position < this.state.drag && dragPosition < midpoint) {
+        newPosition--;
+        return this.state.dragHeight;
+      }
+      if (position > this.state.drag && dragPosition + this.state.dragHeight > midpoint) {
+        newPosition++;
+        return -this.state.dragHeight;
+      }
+      return 0;
+    });
+    this.setState({currentPositions, newPosition});
+  },
+
+  handleDragStop() {
+    if (this.state.drag !== this.state.newPosition) {
+      this.props.reorderLevel(this.props.stage.position, this.state.drag, this.state.newPosition);
+    }
+    this.setState({drag: null, newPosition: null, currentPositions: []});
+    window.removeEventListener('selectstart', this.preventSelect);
+    window.removeEventListener('mousemove', this.handleDrag);
+    window.removeEventListener('mouseup', this.handleDragStop);
+  },
+
+  handleAddLevel() {
+    this.props.addLevel(this.props.stage.position);
+  },
+
+  handleLockableChanged() {
+    const state = this.refs.lockable.checked ? 'lockable' : 'not lockable';
+    console.log(`stage ${this.props.stage.position} marked as ${state}`);
```

Comment:
Also, I suspect that the onChange callback gives us an arg from which we can figure out the target instead of using refs.
i.e.
```
handleLockableChanged(event) {
  const state = event.target.checked ? 'lockable' : ' not lockable';
}